### DISCUSSION TITLE
How to Set Velocity to Zero in Solid Regions Based on Phase Field (ϕ=1) in MOOSE

### DISCUSSION THREAD

**Discussion by mahdipasaie** [idaholab/moose#28950] [Q&A Modules: Phase field] [answered]
Hello,

I'm working on a simulation where part of the domain represents a growing solid region, indicated by ( $\phi = 1 \$) with \( $\phi = -1 \$) representing the liquid phase). As the solid region grows, I need to enforce a condition where the velocity $\( u[\text{qp}] \)$ is set to zero whenever ( $\phi$ = 1 \), while the equations solve normally in regions where \($\phi \neq 1$\).

Ideally, I’d like a conditional setup or loop that checks $\phi$ at each point in the domain and applies the zero-velocity condition dynamically as the solid region expands. Here are my main questions:

1. **Conditionally Enforcing Zero Velocity:** Is there a recommended way to implement a conditional check within MOOSE to set \( $u[\text{qp}] = 0 \)$ whenever \( $\phi = 1 \$)? 

something like: 

```
    if (_phi[_qp]  ==  1)
    {

           make _u[qp] = 0 
    }
    else
    {

        solve the equation
    }
  ```
  
Alternatively, is there a way to dynamically apply boundary conditions or pointwise constraints within the domain during runtime to make the velocity zero wherever $\phi$ is 1? 

Also, I am aware of the penalty method or changing the viscosity field method.

Thank you for your guidance!
Mahdi.

**Comment by GiudGiud** [comment]
Hello

"make _u = 0" with kernels could be done with a penalty approach. What type of variable is u going to be? First order lagrange?

> Alternatively, is there a way to dynamically apply boundary conditions or pointwise constraints within the domain during runtime to make the velocity zero wherever ϕ is 1?

Boundary conditions should be on a surface not a volume. So if you used this, I would set the value at the interface, then make the value constant on the solid domain by only having a NullKernel or a TimeDerivative kernel on the velocity variable in that domain. 

Pointwise constraints could be applied using a penalty approach with a volumetric source pushing velocity to 0
OR
If you create a nodeset for the entire solid domain we could use a NodalKernel to solve an ODE pushing velocity to 0 over the solid domain. 

Changing the viscosity seems like a good idea too. 
There are a few things to try, please report on your success with them.

@lindsayad also created some fluid-solid simulations here that can be an inspiration
https://github.com/idaholab/moose/tree/next/modules/navier_stokes/examples/laser-welding

**Reply by mahdipasaie** [reply]
Thank you! Ideally, I'd like to apply it to the Finite Volume Method (FVM) for velocity field variables. I appreciate the suggestions and will give these methods a try.

**Reply by mahdipasaie** [reply]
After reviewing the finite volume kernels for Navier-Stokes, along with the documentation, I've found the code challenging to interpret, which makes it very difficult for me to implement some of the suggested methods that require kernel modifications.

 As for the variable `mu`or viscosity, doI need to define a `FunctorMaterial` and then pass it to the kernels? 
 
 ```
[./v_viscosity] # μ∇²v
    type = INSFVMomentumDiffusion
    variable = vel_y
    mu = ${mu}
    momentum_component = 'y'
[../]
  
  ```

**Reply by GiudGiud** [reply]
Hello

In this example 'mu' is constant and set to ${mu} which is a constant defined in the input file. But in general yes, 'mu' can be a functor material property, which are defined in FunctorMaterial

**Comment by GiudGiud** [comment]
@mahdipasaie 
I saw you posted an update but cannot find it.

Did you make sure to use AD in that dependence of mu on phi?
Did you get convergence without that dependence but at similar levels of viscosity (same Reynolds number) ?

**Reply by mahdipasaie** [reply]
Hi,

Yes, I was experimenting with something. I used the `mu` functor to make the viscosity dependent on the phase-field order parameter (\(\phi\)) as discussed in this [Nature article](https://www.nature.com/articles/srep45770). The article employs OpenFOAM for the finite volume method (FVM) in fluid flow and the finite element method (FEM) for the phase-field model. While they used a hybrid approach, I reproduced their model entirely using FEM.

Previously, I developed FEM models in FEniCS (segregated approach) to couple the phase-field and Navier-Stokes equations, successfully reproducing results from this [Journal of Computational Physics article](https://www.sciencedirect.com/science/article/pii/S0021999199963234). Below is an example of my implementation:

![Screenshot](https://github.com/user-attachments/assets/855bd3ab-c425-4991-b902-ffb0f594d660)

I found the FVM source code hard to interpret, and the FEM code was complex. As a result, I created my own component-wise FEM Navier-Stokes formulation in MOOSE, which you can see here: [GitHub repository](https://github.com/mahdipasaie/moose_fluid_component.git). I also implemented penalization kernels to enforce fluid exclusion in solid regions for FEM, available here: [GitHub repository](https://github.com/mahdipasaie/pure_coupled_moose.git).

Given my specific interest, I realized I might need a turbulent formulation. While I aim to avoid direct numerical simulation (DNS), turbulence modeling may be necessary if the laminar approach proves insufficient. My tests in OpenFOAM revealed turbulence arising due to complex geometry. I also developed a phase-field FVM model in OpenFOAM, available here: [GitHub repository](https://chatgpt.com/c/git@github.com:mahdipasaie/openfoam_Isothermal.git). However, coupling OpenFOAM’s Isofoam model with my work was non-trivial, so I shifted to MOOSE for its flexibility in coupling FVM and FEM using the MultiApp structure.

To explore this, I tested a simple FVM-FEM coupling example in MOOSE, combining FVM fluid flow with FEM thermal conduction (segregated coupling). I added only a convection term to the heat equation, which worked well: [GitHub repository](https://github.com/mahdipasaie/moose_coupled_heat_fluid.git). 

My ultimate goal was to integrate FVM and FEM for my research, but due to time constraints, I decided to focus solely on FEM to wrap up my thesis. Developing a turbulent model in MOOSE would be highly beneficial for future work. Additionally, using the FVM with the viscous term in my earlier experiments resulted in negative viscosity (`mu`) values. While FEM handled this scenario well, the [Nature article](https://www.nature.com/articles/srep45770) managed it in FVM without issues, so perhaps I made an error in my implementation.

**Reply by GiudGiud** [reply, marked as answer]
Thanks for the details on the project! I see a ton of work has been put into this problem and it is always very encouraging to see you can reproduce prior results. 

> combining FVM fluid flow with FEM thermal conduction (segregated coupling)

This PR will enable direct coupling, but that's likely not the issue
https://github.com/idaholab/moose/pull/29106/

> Additionally, using the FVM with the viscous term in my earlier experiments resulted in negative viscosity (mu) values.

This does not seem right. Is phi properly bounded? and is the viscosity allowed to become negative over the valid range of phi?

**Reply by lindsayad** [reply]
We already have demonstrated two-way FE-FV coupling through the functor system in https://github.com/idaholab/moose/blob/next/modules/navier_stokes/examples/laser-welding/2d-fv.i . We'll have to see how much of #29106 is still relevant post-functor system

**Comment by mahdipasaie** [comment]
Yes, also, Finding a platform to couple FVM and FEM code was quite challenging for me, especially since the source codes were often very complex to understand. In contrast, the FEM kernels in MOOSE are straightforward and much easier to work with. I hope that with the ongoing development of FVM kernels in MOOSE, they will become just as accessible and intuitive as the FEM kernels, making the process easier for others in the future
